File: apps/carbon/lib/api/energy.ts
/**
 * Energy & Scope 2 API client.
 *
 * PR-06A — fondation énergie : compteurs, activités, instruments contractuels
 * et allocations (quantités en MWh, aucun tCO2e).
 * PR-06B — moteur de calcul Scope 2 dual : runs, trace de calcul, approbation,
 * Evidence Pack (`/energy/scope2/*`). C'est la seule partie qui porte des tCO2e,
 * et elle en porte TOUJOURS deux : location-based ET market-based.
 *
 * Types en snake_case, miroir de apps/api/models/energy.py. Réutilise
 * `API_BASE_URL` + `getAuthToken` du client principal (@/lib/api) plutôt que de
 * dupliquer la config.
 */

import { API_BASE_URL, getAuthToken } from "@/lib/api";

// ── Vocabulaires (miroir des Literal Pydantic) ──────────────────────────────
export type Carrier = "electricity" | "gas" | "heat" | "steam" | "cooling" | "other";
export type EnergyDataStatus = "verified" | "estimated" | "manual" | "inferred";
export type EnergyReviewStatus = "pending" | "accepted" | "flagged";
export type InstrumentType = "rec" | "go" | "ppa" | "green_tariff";
export type InstrumentStatus = "active" | "expired" | "cancelled";
export type FactorBasis = "location" | "market" | "residual_mix";

// ── Types de réponse ────────────────────────────────────────────────────────
export interface EnergyMeter {
  id: number;
  company_id: number;
  site_id: number | null;
  carrier: Carrier;
  meter_code: string;
  label: string | null;
  unit: string;
  active: boolean;
  created_at: string;
  updated_at: string;
}

export interface EnergyActivity {
  id: number;
  company_id: number;
  meter_id: number | null;
  site_id: number | null;
  carrier: Carrier;
  quantity: number;
  unit: string;
  period_start: string;
  period_end: string;
  import_id: string | null;
  data_status: EnergyDataStatus;
  evidence_artifact_id: number | null;
  review_status: EnergyReviewStatus;
  created_at: string;
  updated_at: string;
}

export interface ContractualInstrument {
  id: number;
  company_id: number;
  instrument_type: InstrumentType;
  carrier: Carrier;
  reference: string | null;
  volume_mwh: number;
  valid_from: string;
  valid_to: string;
  geography_code: string | null;
  certificate_artifact_id: number | null;
  status: InstrumentStatus;
  created_at: string;
  updated_at: string;
  allocated_mwh: number;
  remaining_mwh: number;
  is_expired: boolean;
}

export interface InstrumentAllocation {
  id: number;
  company_id: number;
  instrument_id: number;
  energy_activity_id: number;
  allocated_mwh: number;
  allocated_at: string;
  allocated_by: number | null;
  created_at: string;
}

export interface ActivityImportResult {
  import_id: string;
  filename: string;
  total_rows: number;
  created: number;
  skipped: number;
  review_status: EnergyReviewStatus;
  warnings: string[];
}

interface Paginated<T> {
  items: T[];
  total: number;
  limit: number;
  offset: number;
}

export type MeterList = Paginated<EnergyMeter>;
export type ActivityList = Paginated<EnergyActivity>;
export type InstrumentList = Paginated<ContractualInstrument>;

// ── Fetch helpers (auth + base URL réutilisés du client principal) ──────────
function authHeaders(): Record<string, string> {
  const token = getAuthToken();
  return token ? { Authorization: `Bearer ${token}` } : {};
}

async function energyGet<T>(path: string, signal?: AbortSignal): Promise<T> {
  const res = await fetch(`${API_BASE_URL}${path}`, {
    method: "GET",
    headers: { Accept: "application/json", ...authHeaders() },
    credentials: "include",
    signal,
  });
  if (!res.ok) throw new Error(`API ${res.status} on ${path}`);
  return (await res.json()) as T;
}

async function energyPost<T>(path: string, body: unknown, signal?: AbortSignal): Promise<T> {
  const res = await fetch(`${API_BASE_URL}${path}`, {
    method: "POST",
    headers: { "Content-Type": "application/json", Accept: "application/json", ...authHeaders() },
    credentials: "include",
    body: JSON.stringify(body),
    signal,
  });
  if (!res.ok) {
    const detail = await res.json().catch(() => null);
    throw new Error(detail?.detail ?? `API ${res.status} on ${path}`);
  }
  return (await res.json()) as T;
}

function qs(params: Record<string, string | number | boolean | undefined>): string {
  const sp = new URLSearchParams();
  for (const [k, v] of Object.entries(params)) {
    if (v !== undefined && v !== "") sp.set(k, String(v));
  }
  const s = sp.toString();
  return s ? `?${s}` : "";
}

// ── Compteurs ───────────────────────────────────────────────────────────────
export function fetchMeters(
  opts: { carrier?: Carrier; activeOnly?: boolean; limit?: number; offset?: number } = {},
  signal?: AbortSignal,
): Promise<MeterList> {
  return energyGet<MeterList>(
    `/energy/meters${qs({ carrier: opts.carrier, active_only: opts.activeOnly, limit: opts.limit, offset: opts.offset })}`,
    signal,
  );
}

export function createMeter(
  body: { carrier: Carrier; meter_code: string; site_id?: number | null; label?: string | null; unit?: string; active?: boolean },
): Promise<EnergyMeter> {
  return energyPost<EnergyMeter>("/energy/meters", body);
}

// ── Activités ─────────────────────────────────────────────────────────────
export function fetchActivities(
  opts: {
    site_id?: number; carrier?: Carrier; review_status?: EnergyReviewStatus;
    period_from?: string; period_to?: string; limit?: number; offset?: number;
  } = {},
  signal?: AbortSignal,
): Promise<ActivityList> {
  return energyGet<ActivityList>(`/energy/activities${qs(opts)}`, signal);
}

export function importActivities(filename: string, csvText: string): Promise<ActivityImportResult> {
  return energyPost<ActivityImportResult>("/energy/activities/import", { filename, csv_text: csvText });
}

// ── Instruments & allocations ─────────────────────────────────────────────
export function fetchInstruments(
  opts: { carrier?: Carrier; status?: InstrumentStatus; limit?: number; offset?: number } = {},
  signal?: AbortSignal,
): Promise<InstrumentList> {
  return energyGet<InstrumentList>(`/energy/instruments${qs(opts)}`, signal);
}

export function createInstrument(body: {
  instrument_type: InstrumentType; volume_mwh: number; valid_from: string; valid_to: string;
  carrier?: Carrier; reference?: string | null; geography_code?: string | null; certificate_artifact_id?: number | null;
}): Promise<ContractualInstrument> {
  return energyPost<ContractualInstrument>("/energy/instruments", body);
}

export function allocateInstrument(
  instrumentId: number,
  body: { energy_activity_id: number; allocated_mwh: number },
): Promise<InstrumentAllocation> {
  return energyPost<InstrumentAllocation>(`/energy/instruments/${instrumentId}/allocate`, body);
}

// ══════════════════════════════════════════════════════════════════════════
// Moteur de calcul Scope 2 dual (PR-06B)
// ══════════════════════════════════════════════════════════════════════════

export type Scope2Basis = "location" | "market";
export type Scope2Segment = "total" | "covered" | "uncovered";
export type Scope2RunStatus = "draft" | "approved" | "superseded";
export type Scope2FactorBasis =
  | "location"
  | "market"
  | "residual_mix"
  | "contractual_instrument"
  | "documented_fallback";

/** Enveloppe analytique `{data, meta, evidence}` (contrats Wave 2 §4) : un
 *  résultat n'arrive jamais nu, toujours avec sa date, sa méthode versionnée,
 *  son statut et sa qualité. */
export interface AnalyticalEnvelope<T> {
  data: T;
  meta: {
    as_of: string;
    status: EnergyDataStatus;
    method: { code: string; version: string };
    quality: { confidence: number | null; coverage_pct: number | null; warnings: string[] };
  };
  evidence: Array<{
    artifact_id: number | null;
    source_code: string | null;
    release_key: string | null;
    page_reference: string | null;
    note: string | null;
  }>;
}

/** Une ligne de Trace de calcul. `selection_level` et `selection_reason` sont
 *  toujours renseignés — aucun facteur n'est choisi silencieusement. */
export interface Scope2TraceLine {
  id: number | null;
  energy_activity_id: number | null;
  basis: Scope2Basis;
  segment: Scope2Segment;
  instrument_id: number | null;
  carrier: Carrier;
  geography_code: string | null;
  period_start: string;
  period_end: string;
  activity_value: number;
  activity_unit: string;
  activity_mwh: number;
  ef_id: number | null;
  ef_code: string | null;
  ef_version: string | null;
  factor_kgco2e_per_mwh: number | null;
  factor_basis: Scope2FactorBasis | null;
  selection_level: string;
  selection_reason: string;
  result_tco2e: number;
  uncertainty: number | null;
  data_quality: EnergyDataStatus;
  fallback_reason: string | null;
  warnings: string[];
}

export interface Scope2MissingFactor {
  energy_activity_id: number;
  basis: Scope2Basis;
  segment: Scope2Segment;
  carrier: Carrier;
  geography_code: string | null;
  activity_mwh: number;
  message: string;
}

export interface Scope2ResultData {
  run_id: number;
  status: Scope2RunStatus;
  period_start: string;
  period_end: string;
  geography_code: string;
  location_based_tco2e: number;
  market_based_tco2e: number;
  total_consumption_mwh: number;
  calculated_consumption_mwh: number;
  contractual_coverage_mwh: number;
  contractual_coverage_pct: number;
  uncovered_mwh: number;
  residual_mix_used: boolean;
  is_complete: boolean;
  input_fingerprint: string;
  calculated_at: string;
  approved_at: string | null;
  missing_factors: Scope2MissingFactor[];
  factors_used: Array<{
    ef_id: number;
    ef_code: string | null;
    ef_version: string | null;
    basis: string;
    selection_level: string;
    factor_basis: string;
  }>;
  trace: Scope2TraceLine[];
}

export interface Scope2RunSummary {
  id: number;
  company_id: number;
  methodology_code: string;
  methodology_version: string;
  period_start: string;
  period_end: string;
  geography_code: string;
  status: Scope2RunStatus;
  location_based_tco2e: number | null;
  market_based_tco2e: number | null;
  confidence: number | null;
  coverage_pct: number | null;
  is_complete: boolean;
  input_fingerprint: string;
  calculated_at: string;
  approved_at: string | null;
  warning_count: number;
}

export type Scope2RunEnvelope = AnalyticalEnvelope<Scope2ResultData>;
export type Scope2RunList = Paginated<Scope2RunSummary>;

export function fetchScope2Runs(
  opts: { status?: Scope2RunStatus; limit?: number; offset?: number } = {},
  signal?: AbortSignal,
): Promise<Scope2RunList> {
  return energyGet<Scope2RunList>(`/energy/scope2/runs${qs(opts)}`, signal);
}

export function fetchScope2Run(runId: number, signal?: AbortSignal): Promise<Scope2RunEnvelope> {
  return energyGet<Scope2RunEnvelope>(`/energy/scope2/runs/${runId}`, signal);
}

export function fetchScope2Trace(
  runId: number,
  signal?: AbortSignal,
): Promise<{ run_id: number; items: Scope2TraceLine[]; total: number }> {
  return energyGet(`/energy/scope2/runs/${runId}/trace`, signal);
}

export function calculateScope2(body: {
  period_start: string;
  period_end: string;
  geography_code: string;
  site_geographies?: Record<number, string>;
  include_pending?: boolean;
  /** Niveau 4 de la hiérarchie market-based. Faux par défaut : sans
   *  autorisation méthodologique explicite, l'absence de facteur de marché est
   *  une erreur remontée, jamais un repli silencieux. */
  allow_market_fallback?: boolean;
  fallback_note?: string | null;
}): Promise<Scope2RunEnvelope> {
  return energyPost<Scope2RunEnvelope>("/energy/scope2/calculate", body);
}

export function approveScope2Run(runId: number): Promise<Scope2RunEnvelope> {
  return energyPost<Scope2RunEnvelope>(`/energy/scope2/runs/${runId}/approve`, {});
}

/** URL de téléchargement de l'Evidence Pack (ZIP signé). Le téléchargement passe
 *  par une requête authentifiée — pas d'URL signée permanente (contrats §3). */
export async function downloadScope2EvidencePack(runId: number): Promise<Blob> {
  const res = await fetch(`${API_BASE_URL}/energy/scope2/runs/${runId}/evidence-pack`, {
    method: "GET",
    headers: { ...authHeaders() },
    credentials: "include",
  });
  if (!res.ok) throw new Error(`API ${res.status} on evidence-pack`);
  return res.blob();
}

File: apps/api/models/energy.py
"""
models/energy.py — Énergie & Scope 2 (PR-06A) : compteurs, activités,
instruments contractuels, allocations, métadonnées de facteurs.

Un modèle par table + un modèle de requête par opération d'écriture exposée.
Champs snake_case, alignés sur les colonnes SQL de la migration 031 (même
convention que models/intelligence.py — API interne). Aucun total Scope 2,
aucun résultat de calcul ici : PR-06A est une FONDATION de données, le moteur
de calcul dual LB/MB est PR-06B.
"""

from __future__ import annotations

from datetime import date, datetime
from typing import Literal

from pydantic import BaseModel, Field

Carrier = Literal["electricity", "gas", "heat", "steam", "cooling", "other"]
EnergyDataStatus = Literal["verified", "estimated", "manual", "inferred"]
ReviewStatus = Literal["pending", "accepted", "flagged"]
InstrumentType = Literal["rec", "go", "ppa", "green_tariff"]
InstrumentStatus = Literal["active", "expired", "cancelled"]
FactorBasis = Literal["location", "market", "residual_mix"]


# ---------------------------------------------------------------------------
# energy_meters
# ---------------------------------------------------------------------------

class MeterCreate(BaseModel):
    carrier: Carrier
    meter_code: str = Field(min_length=1)
    site_id: int | None = None
    label: str | None = None
    unit: str = "MWh"
    active: bool = True


class MeterResponse(BaseModel):
    id: int
    company_id: int
    site_id: int | None
    carrier: Carrier
    meter_code: str
    label: str | None
    unit: str
    active: bool
    created_at: datetime
    updated_at: datetime


class MeterListResponse(BaseModel):
    items: list[MeterResponse]
    total: int
    limit: int
    offset: int


# ---------------------------------------------------------------------------
# energy_activities
# ---------------------------------------------------------------------------

class ActivityImportRequest(BaseModel):
    """Import CSV d'activités énergie. `csv_text` = contenu brut du fichier
    (colonnes : meter_code, carrier, quantity, unit, period_start, period_end).

    Le front lit le fichier et poste son texte — pas de parseur multipart
    bespoke. Idempotent : réimporter le même CSV ne crée aucun doublon (clé
    naturelle company_id+meter+période UNIQUE en base)."""
    filename: str = Field(min_length=1)
    csv_text: str


class ActivityImportResult(BaseModel):
    import_id: str
    filename: str
    total_rows: int
    created: int
    skipped: int
    review_status: ReviewStatus = "pending"
    warnings: list[str] = Field(default_factory=list)


class ActivityResponse(BaseModel):
    id: int
    company_id: int
    meter_id: int | None
    site_id: int | None
    carrier: Carrier
    quantity: float
    unit: str
    period_start: date
    period_end: date
    import_id: str | None
    data_status: EnergyDataStatus
    evidence_artifact_id: int | None
    review_status: ReviewStatus
    created_at: datetime
    updated_at: datetime


class ActivityListResponse(BaseModel):
    items: list[ActivityResponse]
    total: int
    limit: int
    offset: int


# ---------------------------------------------------------------------------
# contractual_instruments
# ---------------------------------------------------------------------------

class InstrumentCreate(BaseModel):
    instrument_type: InstrumentType
    volume_mwh: float = Field(ge=0)
    valid_from: date
    valid_to: date
    carrier: Carrier = "electricity"
    reference: str | None = None
    geography_code: str | None = None
    certificate_artifact_id: int | None = None


class InstrumentResponse(BaseModel):
    id: int
    company_id: int
    instrument_type: InstrumentType
    carrier: Carrier
    reference: str | None
    volume_mwh: float
    valid_from: date
    valid_to: date
    geography_code: str | None
    certificate_artifact_id: int | None
    status: InstrumentStatus
    created_at: datetime
    updated_at: datetime
    # Champs dérivés (couverture / expiry) — calculés à la lecture, pas stockés.
    allocated_mwh: float = 0.0
    remaining_mwh: float = 0.0
    is_expired: bool = False


class InstrumentListResponse(BaseModel):
    items: list[InstrumentResponse]
    total: int
    limit: int
    offset: int


# ---------------------------------------------------------------------------
# instrument_allocations
# ---------------------------------------------------------------------------

class AllocationRequest(BaseModel):
    energy_activity_id: int
    allocated_mwh: float = Field(gt=0)


class AllocationResponse(BaseModel):
    id: int
    company_id: int
    instrument_id: int
    energy_activity_id: int
    allocated_mwh: float
    allocated_at: datetime
    allocated_by: int | None


# ---------------------------------------------------------------------------
# Moteur de calcul Scope 2 dual (PR-06B) — scope2_calculation_runs /
# scope2_line_results. Ici, et seulement ici, apparaissent des tCO2e.
# ---------------------------------------------------------------------------

Scope2Basis = Literal["location", "market"]
Scope2Segment = Literal["total", "covered", "uncovered"]
Scope2FactorBasis = Literal[
    "location", "market", "residual_mix", "contractual_instrument", "documented_fallback"
]
Scope2RunStatus = Literal["draft", "approved", "superseded"]


class Scope2CalculateRequest(BaseModel):
    """Demande de calcul. `geography_code` (zone de réseau, ex. `FR` ou
    `FR-IDF`) est OBLIGATOIRE : un calcul location-based sans zone déclarée
    n'existe pas, et deviner la zone serait un choix silencieux.

    `allow_market_fallback` est le seul interrupteur autorisant le niveau 4 de
    la hiérarchie market-based ; **faux par défaut** — sans autorisation
    méthodologique explicite, l'absence de facteur de marché est une erreur
    remontée, jamais un repli discret sur une moyenne de réseau.
    """

    period_start: date
    period_end: date
    geography_code: str = Field(min_length=2)
    site_geographies: dict[int, str] = Field(default_factory=dict)
    include_pending: bool = True
    allow_market_fallback: bool = False
    fallback_note: str | None = None


class Scope2MissingFactor(BaseModel):
    energy_activity_id: int
    basis: Scope2Basis
    segment: Scope2Segment
    carrier: Carrier
    geography_code: str | None
    activity_mwh: float
    message: str


class Scope2FactorUsed(BaseModel):
    ef_id: int
    ef_code: str | None = None
    ef_version: str | None = None
    basis: str
    selection_level: str
    factor_basis: str


class Scope2TraceLine(BaseModel):
    """Une ligne de la Trace de calcul. `selection_level` + `selection_reason`
    sont NON NULS : la trace dit toujours d'où vient le facteur retenu."""

    id: int | None = None
    energy_activity_id: int | None
    basis: Scope2Basis
    segment: Scope2Segment
    instrument_id: int | None = None
    carrier: Carrier
    geography_code: str | None
    period_start: date
    period_end: date
    activity_value: float
    activity_unit: str
    activity_mwh: float
    ef_id: int | None = None
    ef_code: str | None = None
    ef_version: str | None = None
    factor_kgco2e_per_mwh: float | None = None
    factor_basis: Scope2FactorBasis | None = None
    selection_level: str
    selection_reason: str
    result_tco2e: float
    uncertainty: float | None = None
    data_quality: EnergyDataStatus
    fallback_reason: str | None = None
    warnings: list[str] = Field(default_factory=list)


class Scope2ResultData(BaseModel):
    """Le `data` de l'enveloppe analytique (contrats §4).

    Les DEUX totaux coexistent toujours — aucun champ ne peut en masquer un
    autre côté API, comme côté UI.
    """

    run_id: int
    status: Scope2RunStatus
    period_start: date
    period_end: date
    geography_code: str
    location_based_tco2e: float
    market_based_tco2e: float
    total_consumption_mwh: float
    calculated_consumption_mwh: float
    contractual_coverage_mwh: float
    contractual_coverage_pct: float
    uncovered_mwh: float
    residual_mix_used: bool
    is_complete: bool
    input_fingerprint: str
    calculated_at: datetime
    approved_at: datetime | None = None
    missing_factors: list[Scope2MissingFactor] = Field(default_factory=list)
    factors_used: list[Scope2FactorUsed] = Field(default_factory=list)
    trace: list[Scope2TraceLine] = Field(default_factory=list)


class Scope2RunSummary(BaseModel):
    """Vue de liste d'un run — sans la trace complète (payload contenu)."""

    id: int
    company_id: int
    methodology_code: str
    methodology_version: str
    period_start: date
    period_end: date
    geography_code: str
    status: Scope2RunStatus
    location_based_tco2e: float | None = None
    market_based_tco2e: float | None = None
    confidence: float | None = None
    coverage_pct: float | None = None
    is_complete: bool = False
    input_fingerprint: str
    calculated_at: datetime
    approved_at: datetime | None = None
    warning_count: int = 0


class Scope2RunListResponse(BaseModel):
    items: list[Scope2RunSummary]
    total: int
    limit: int
    offset: int


class Scope2TraceResponse(BaseModel):
    run_id: int
    items: list[Scope2TraceLine]
    total: int
    created_at: datetime
